File: device/metal/elementwise/native/axpy.m
#include "axpy.h"
#include "elementwise.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

int metal_dispatch_axpy(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef yRef,
    MetalBufferRef xRef,
    float alpha,
    uint32_t count,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_elementwise_status_clear(status);

        if (count == 0) {
            return 0;
        }

        if (yRef == NULL || xRef == NULL) {
            metal_elementwise_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        char kernelName[128];
        int nameCode = metal_elementwise_compose_kernel_name(
            kernelName,
            sizeof(kernelName),
            "axpy",
            metal_elementwise_element_dtype_suffix(elementDType),
            status
        );

        if (nameCode != 0) {
            return nameCode;
        }

        MetalContext* context = (MetalContext*)contextRef;

        if (context == NULL || context->queue == NULL) {
            metal_elementwise_status_set(status, -1, "invalid Metal context");
            return -1;
        }

        id<MTLComputePipelineState> pipeline = metal_get_pipeline(context, kernelName, status);

        if (pipeline == nil) {
            return status != NULL && status->code != 0 ? status->code : -7;
        }

        id<MTLCommandBuffer> commandBuffer;
        id<MTLComputeCommandEncoder> encoder = metal_get_encoder(context, &commandBuffer);

        [encoder setComputePipelineState:pipeline];
        [encoder setBuffer:(__bridge id<MTLBuffer>)yRef offset:0 atIndex:0];
        [encoder setBuffer:(__bridge id<MTLBuffer>)xRef offset:0 atIndex:1];
        [encoder setBytes:&alpha length:sizeof(alpha) atIndex:2];
        [encoder setBytes:&count length:sizeof(count) atIndex:3];

        NSUInteger threadWidth = [pipeline threadExecutionWidth];

        if (threadWidth == 0) {
            threadWidth = 1;
        }

        NSUInteger vectorCount = (NSUInteger)((count + 3) / 4);
        [encoder
            dispatchThreads:MTLSizeMake(vectorCount, 1, 1)
            threadsPerThreadgroup:MTLSizeMake(threadWidth, 1, 1)
        ];
        metal_track_command_completion(context, commandBuffer, completionToken, NULL);
        metal_end_encoder(context, encoder, commandBuffer);

        return 0;
    }
}

File: device/cuda/elementwise/native/axpy.c
#include "axpy.h"
#include "elementwise.h"
#include "../internal/bridge/core_private.h"

#include <stdint.h>

int cuda_dispatch_axpy(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef yRef,
    CUDABufferRef xRef,
    float alpha,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_elementwise_status_clear(status);

    if (count == 0) {
        return 0;
    }

    if (yRef == NULL || xRef == NULL) {
        cuda_elementwise_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    const char* dtypeSuffix = cuda_elementwise_element_dtype_suffix(elementDType);

    if (dtypeSuffix == NULL) {
        cuda_elementwise_status_set(status, -6, "unknown axpy dtype");
        return -6;
    }

    char kernelName[128];
    int nameCode = cuda_elementwise_compose_kernel_name(
        kernelName,
        sizeof(kernelName),
        "axpy",
        dtypeSuffix,
        status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_elementwise_module_source();

    if (moduleSource == NULL) {
        cuda_elementwise_status_set(status, -7, "CUDA elementwise module source not registered");
        return -7;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    void* yPtr = cuda_buffer_device_ptr(yRef);
    void* xPtr = cuda_buffer_device_ptr(xRef);
    void* args[] = {&yPtr, &xPtr, &count, &alpha};
    int launchCode = cuda_launch_1d(context, kernel, stream, count, args, sizeof(args), status);

    if (launchCode != 0) {
        return launchCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}
